src/providers/hana/qgshanafeatureiterator.cpp

```diff
@@ -0,0 +1,437 @@
+    sqlFilter = getBBOXFilter( QgsHanaUtils::toHANAVersion( mConnection->getDatabaseVersion() ) );
+
+  if ( !mSource->mQueryWhereClause.isEmpty() )
+    sqlFilter = andWhereClauses( sqlFilter, mSource->mQueryWhereClause );
+
+  // Set fid filter
+  if ( !mFidColumn.isEmpty() )
+  {
+    if ( request.filterType() == QgsFeatureRequest::FilterFid )
+    {
+      QString inClause = QStringLiteral( " %1 = %2" ).arg(
+                           QgsHanaUtils::quotedIdentifier( mFidColumn ), FID_TO_STRING( request.filterFid() ) );
+      sqlFilter = andWhereClauses( sqlFilter, inClause );
+    }
+    else if ( request.filterType() == QgsFeatureRequest::FilterFids && !mRequest.filterFids().isEmpty() )
+    {
+      QString delim;
+      QString inClause = QStringLiteral( "%1 IN (" ).arg( QgsHanaUtils::quotedIdentifier( mFidColumn ) );
+      Q_FOREACH ( QgsFeatureId featureId, mRequest.filterFids() )
+      {
+        inClause += delim + FID_TO_STRING( featureId );
+        delim = ',';
+      }
+      inClause.append( ')' );
+
+      sqlFilter = andWhereClauses( sqlFilter, inClause );
+    }
+  }
+
+  //IMPORTANT - this MUST be the last clause added
+  mExpressionCompiled = false;
+  mCompileStatus = NoCompilation;
+  if ( request.filterType() == QgsFeatureRequest::FilterExpression )
+  {
+    if ( QgsSettings().value( QStringLiteral( "qgis/compileExpressions" ), true ).toBool() )
+    {
+      QgsHanaExpressionCompiler compiler = QgsHanaExpressionCompiler( mSource );
+      QgsSqlExpressionCompiler::Result result = compiler.compile( request.filterExpression() );
+      switch ( result )
+      {
+        case QgsSqlExpressionCompiler::Result::Complete:
+        case QgsSqlExpressionCompiler::Result::Partial:
+        {
+          QString filterExpr = compiler.result();
+          if ( !filterExpr.isEmpty() )
+          {
+            sqlFilter = andWhereClauses( sqlFilter, filterExpr );
+            //if only partial success when compiling expression, we need to double-check results
+            //using QGIS' expressions
+            mExpressionCompiled = ( result == QgsSqlExpressionCompiler::Result::Complete );
+            mCompileStatus = ( mExpressionCompiled ? Compiled : PartiallyCompiled );
+          }
+        }
+        break;
+        case QgsSqlExpressionCompiler::Result::Fail:
+          QgsDebugMsg( QStringLiteral( "Unable to compile filter expression: '%1'" )
+                       .arg( request.filterExpression()->expression() ).toStdString().c_str() );
+          break;
+        case QgsSqlExpressionCompiler::Result::None:
+          break;
+      }
+      if ( result != QgsSqlExpressionCompiler::Result::Complete )
+      {
+        //can't apply limit at provider side as we need to check all results using QGIS expressions
+        limitAtProvider = false;
+      }
+    }
+    else
+    {
+      limitAtProvider = false;
+    }
+  }
+
+  if ( !sqlFilter.isEmpty() )
+    sql += QStringLiteral( " WHERE " ) + sqlFilter;
```

Comment:
I'd also propose to use `QStringList` for `sqlFilter` and to collect all expressions, which are then combinded via `AND`:
```c++
if ( !sqlFilter.isEmpty() )
  sql += QStringLiteral( " WHERE (" ) + sqlFilter.join( ") AND (" ) + QStringLiteral( ")" );
```

This saves quite some parentheses. `(((A) AND (B)) AND (C)) AND (D)` becomes just `(A) AND (B) AND (C) AND (D)`.

Reply:
Done in 38bdec5.